frontend/src/ui/multi_select/multi_select.tsx

```diff
@@ -0,0 +1,121 @@
+            tempSelectedCodes: []
+        })
+    }, [allItems, initialSelection]);
+
+    const onItemClick = (item: T, stateProperty: string) => {
+        const data: Array<string> = [...state[stateProperty]];
+        const index = data.indexOf(item.code);
+
+        index === -1 ? data.push(item.code) : data.splice(index, 1);
+
+        setState({...state, [stateProperty]: data})
+    };
+
+    const onItemRender = (item: T, stateProperty: string) => {
+        const data: Array<string> = state[stateProperty];
+        const isSelected = data.includes(item.code);
+
+        return (
+            <label htmlFor={item.code}
+                   className={cls( isSelected && css.selected)}>
+                <input type="checkbox" name={item.code} id={item.code} checked={isSelected}/>
+                {itemRender(item)}
+            </label>
+        )
+    };
+
+    //todo remove duplication :(
+
+    const onAdd = () => {
+        const _itemsAvailable = [...itemsAvailable].filter(i => !tempAvailableCodes.includes(i.code));
+        const _itemsSelected = [...itemsSelected, ...tempAvailableCodes.map(i => allItems.find(j => j.code === i))];
+        const _tempSelectedCodes = [...tempAvailableCodes];
+
+        _onChange(_itemsSelected);
+        setState({itemsSelected: _itemsSelected, itemsAvailable: _itemsAvailable, tempSelectedCodes: _tempSelectedCodes, tempAvailableCodes: []})
+    };
+
+    const onAddAll = () => {
+        const _itemsAvailable = [];
+        const _itemsSelected = [...allItems];
+        const _tempSelectedCodes = [...tempAvailableCodes];
+
+        _onChange(_itemsSelected);
+        setState({itemsSelected: _itemsSelected, itemsAvailable: _itemsAvailable, tempSelectedCodes: _tempSelectedCodes, tempAvailableCodes: []})
+    };
+
+    const onRemoveAll = () => {
+        const _itemsSelected = [];
+        const _itemsAvailable = [...allItems];
+        const _tempAvailableCodes = [...tempSelectedCodes];
+
+        _onChange(_itemsSelected);
+        setState({...state, itemsSelected: _itemsSelected, itemsAvailable: _itemsAvailable, tempAvailableCodes: _tempAvailableCodes, tempSelectedCodes: []})
+    };
+
+    const onRemove = () => {
+        const _itemsSelected = [...itemsSelected].filter(i => !tempSelectedCodes.includes(i.code));
+        const _itemsAvailable = [...itemsAvailable, ...tempSelectedCodes.map(i => allItems.find(j => j.code === i))];
+        const _tempAvailableCodes = [...tempSelectedCodes];
+
+        _onChange(_itemsSelected);
+        setState({itemsSelected: _itemsSelected, itemsAvailable: _itemsAvailable, tempAvailableCodes: _tempAvailableCodes, tempSelectedCodes: []})
+    };
+
+    const _onChange = (itemsSelected: Array<T>) => {
+        const added = itemsSelected.filter(i => !initialSelection.some(j => j.code === i.code));
+        const removed = initialSelection.filter(i => !itemsSelected.some(j => j.code === i.code));
+
+        onChange({added, removed})
+    };
```

Comment:
useCallback, or class component.
hooks are nice when they improve readability.
here it's not very readable.

you can use hooks and class components in the same project, no biggie.